ast/src/parsed/pattern.rs

```diff
@@ -0,0 +1,358 @@
+use powdr_number::BigInt;
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use std::iter::{empty, once};
+
+use super::{asm::SymbolPath, visitor::Children};
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Serialize, Deserialize, JsonSchema)]
+pub enum Pattern {
+    CatchAll, // "_", matches a single value
+    Ellipsis, // "..", matches a series of values, only valid inside array patterns
+    #[schemars(skip)]
+    Number(BigInt),
+    String(String),
+    Tuple(Vec<Pattern>),
+    Array(Vec<Pattern>),
+    // A pattern that binds a variable. Variable references are parsed as
+    // Enum and are then re-mapped to Variable if they do not reference
+    // an enum variant.
+    Variable(String),
+    Enum(SymbolPath, Option<Vec<Pattern>>),
+}
+
+impl Pattern {
+    /// Returns an iterator over all variables in this pattern.
+    pub fn variables(&self) -> Box<dyn Iterator<Item = &String> + '_> {
+        match self {
+            Pattern::Variable(v) => Box::new(once(v)),
+            _ => Box::new(self.children().flat_map(|p| p.variables())),
+        }
+    }
+
+    /// Return true if the pattern is irrefutable, i.e. matches all possible values of its type.
+    pub fn is_irrefutable(&self) -> bool {
+        match self {
+            Pattern::Ellipsis => unreachable!(),
+            Pattern::CatchAll | Pattern::Variable(_) => true,
+            Pattern::Number(_) | Pattern::String(_) | Pattern::Enum(_, _) => false,
+            Pattern::Array(items) => {
+                // Only "[..]"" is irrefutable
+                items == &vec![Pattern::Ellipsis]
+            }
+            Pattern::Tuple(p) => p.iter().all(|p| p.is_irrefutable()),
+        }
+    }
+
+    // Specialize a pattern based on a "constructor" pattern passed as a parameter.
```

Comment:
Can you use triple-slash and explain at least in some words what this does, so one can get a general idea without having to open that link?